From 6455618bfaaf111df4e392b37b2d4d6c0636f851 Mon Sep 17 00:00:00 2001
From: kk <kaiwalya.khasnis57@gmail.com>
Date: Wed, 14 Oct 2026 11:33:30 +0900
Subject: [PATCH] rgbled: add vendor.lineage.rgbled@1.0 HIDL interface

IRgbLed gives direct control of the notification RGB LED: per-channel
brightness, solid color, hardware blink and kernel LED triggers.
---
 rgbled/1.0/Android.bp  |  89 +++++++++++++++++++++++
 rgbled/1.0/Android.mk  | 156 +++++++++++++++++++++++++++++++++++++++++
 rgbled/1.0/IRgbLed.hal |  48 +++++++++++++
 rgbled/1.0/types.hal   |  31 ++++++++
 rgbled/Android.bp      |   4 ++
 5 files changed, 328 insertions(+)
 create mode 100644 rgbled/1.0/Android.bp
 create mode 100644 rgbled/1.0/Android.mk
 create mode 100644 rgbled/1.0/IRgbLed.hal
 create mode 100644 rgbled/1.0/types.hal
 create mode 100644 rgbled/Android.bp

diff --git a/rgbled/1.0/Android.bp b/rgbled/1.0/Android.bp
new file mode 100644
index 0000000..bc12dcb
--- /dev/null
+++ b/rgbled/1.0/Android.bp
@@ -0,0 +1,89 @@
+// This file is autogenerated by hidl-gen. Do not edit manually.
+
+filegroup {
+    name: "vendor.lineage.rgbled@1.0_hal",
+    srcs: [
+        "types.hal",
+        "IRgbLed.hal",
+    ],
+}
+
+genrule {
+    name: "vendor.lineage.rgbled@1.0_genc++",
+    tools: ["hidl-gen"],
+    cmd: "$(location hidl-gen) -o $(genDir) -Lc++-sources -randroid.hidl:system/libhidl/transport -rvendor.lineage:hardware/lineage/interfaces vendor.lineage.rgbled@1.0",
+    srcs: [
+        ":vendor.lineage.rgbled@1.0_hal",
+    ],
+    out: [
+        "vendor/lineage/rgbled/1.0/types.cpp",
+        "vendor/lineage/rgbled/1.0/RgbLedAll.cpp",
+    ],
+}
+
+genrule {
+    name: "vendor.lineage.rgbled@1.0_genc++_headers",
+    tools: ["hidl-gen"],
+    cmd: "$(location hidl-gen) -o $(genDir) -Lc++-headers -randroid.hidl:system/libhidl/transport -rvendor.lineage:hardware/lineage/interfaces vendor.lineage.rgbled@1.0",
+    srcs: [
+        ":vendor.lineage.rgbled@1.0_hal",
+    ],
+    out: [
+        "vendor/lineage/rgbled/1.0/types.h",
+        "vendor/lineage/rgbled/1.0/hwtypes.h",
+        "vendor/lineage/rgbled/1.0/IRgbLed.h",
+        "vendor/lineage/rgbled/1.0/IHwRgbLed.h",
+        "vendor/lineage/rgbled/1.0/BnHwRgbLed.h",
+        "vendor/lineage/rgbled/1.0/BpHwRgbLed.h",
+        "vendor/lineage/rgbled/1.0/BsRgbLed.h",
+    ],
+}
+
+cc_library {
+    name: "vendor.lineage.rgbled@1.0",
+    defaults: ["hidl-module-defaults"],
+    generated_sources: ["vendor.lineage.rgbled@1.0_genc++"],
+    generated_headers: ["vendor.lineage.rgbled@1.0_genc++_headers"],
+    export_generated_headers: ["vendor.lineage.rgbled@1.0_genc++_headers"],
+    vendor_available: true,
+    vndk: {
+        enabled: true,
+    },
+    shared_libs: [
+        "libhidlbase",
+        "libhidltransport",
+        "libhwbinder",
+        "liblog",
+        "libutils",
+        "libcutils",
+    ],
+    export_shared_lib_headers: [
+        "libhidlbase",
+        "libhidltransport",
+        "libhwbinder",
+        "libutils",
+    ],
+}
+
+cc_library {
+    name: "vendor.lineage.rgbled@1.0_vendor",
+    defaults: ["hidl-module-defaults"],
+    generated_sources: ["vendor.lineage.rgbled@1.0_genc++"],
+    generated_headers: ["vendor.lineage.rgbled@1.0_genc++_headers"],
+    export_generated_headers: ["vendor.lineage.rgbled@1.0_genc++_headers"],
+    vendor: true,
+    shared_libs: [
+        "libhidlbase",
+        "libhidltransport",
+        "libhwbinder",
+        "liblog",
+        "libutils",
+        "libcutils",
+    ],
+    export_shared_lib_headers: [
+        "libhidlbase",
+        "libhidltransport",
+        "libhwbinder",
+        "libutils",
+    ],
+}
diff --git a/rgbled/1.0/Android.mk b/rgbled/1.0/Android.mk
new file mode 100644
index 0000000..16094c9
--- /dev/null
+++ b/rgbled/1.0/Android.mk
@@ -0,0 +1,156 @@
+# This file is autogenerated by hidl-gen. Do not edit manually.
+
+LOCAL_PATH := $(call my-dir)
+
+################################################################################
+
+include $(CLEAR_VARS)
+LOCAL_MODULE := vendor.lineage.rgbled-V1.0-java
+LOCAL_MODULE_CLASS := JAVA_LIBRARIES
+
+intermediates := $(call local-generated-sources-dir, COMMON)
+
+HIDL := $(HOST_OUT_EXECUTABLES)/hidl-gen$(HOST_EXECUTABLE_SUFFIX)
+
+LOCAL_JAVA_LIBRARIES := \
+    android.hidl.base-V1.0-java \
+
+
+#
+# Build types.hal (Channel)
+#
+GEN := $(intermediates)/vendor/lineage/rgbled/V1_0/Channel.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.rgbled@1.0::types.Channel
+
+$(GEN): $(LOCAL_PATH)/types.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+
+#
+# Build types.hal (Status)
+#
+GEN := $(intermediates)/vendor/lineage/rgbled/V1_0/Status.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.rgbled@1.0::types.Status
+
+$(GEN): $(LOCAL_PATH)/types.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+
+#
+# Build IRgbLed.hal
+#
+GEN := $(intermediates)/vendor/lineage/rgbled/V1_0/IRgbLed.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/IRgbLed.hal
+$(GEN): PRIVATE_DEPS += $(LOCAL_PATH)/types.hal
+$(GEN): $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.rgbled@1.0::IRgbLed
+
+$(GEN): $(LOCAL_PATH)/IRgbLed.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+include $(BUILD_JAVA_LIBRARY)
+
+
+################################################################################
+
+include $(CLEAR_VARS)
+LOCAL_MODULE := vendor.lineage.rgbled-V1.0-java-static
+LOCAL_MODULE_CLASS := JAVA_LIBRARIES
+
+intermediates := $(call local-generated-sources-dir, COMMON)
+
+HIDL := $(HOST_OUT_EXECUTABLES)/hidl-gen$(HOST_EXECUTABLE_SUFFIX)
+
+LOCAL_STATIC_JAVA_LIBRARIES := \
+    android.hidl.base-V1.0-java-static \
+
+
+#
+# Build types.hal (Channel)
+#
+GEN := $(intermediates)/vendor/lineage/rgbled/V1_0/Channel.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.rgbled@1.0::types.Channel
+
+$(GEN): $(LOCAL_PATH)/types.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+
+#
+# Build types.hal (Status)
+#
+GEN := $(intermediates)/vendor/lineage/rgbled/V1_0/Status.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.rgbled@1.0::types.Status
+
+$(GEN): $(LOCAL_PATH)/types.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+
+#
+# Build IRgbLed.hal
+#
+GEN := $(intermediates)/vendor/lineage/rgbled/V1_0/IRgbLed.java
+$(GEN): $(HIDL)
+$(GEN): PRIVATE_HIDL := $(HIDL)
+$(GEN): PRIVATE_DEPS := $(LOCAL_PATH)/IRgbLed.hal
+$(GEN): PRIVATE_DEPS += $(LOCAL_PATH)/types.hal
+$(GEN): $(LOCAL_PATH)/types.hal
+$(GEN): PRIVATE_OUTPUT_DIR := $(intermediates)
+$(GEN): PRIVATE_CUSTOM_TOOL = \
+        $(PRIVATE_HIDL) -o $(PRIVATE_OUTPUT_DIR) \
+        -Ljava \
+        -randroid.hidl:system/libhidl/transport \
+        -rvendor.lineage:hardware/lineage/interfaces \
+        vendor.lineage.rgbled@1.0::IRgbLed
+
+$(GEN): $(LOCAL_PATH)/IRgbLed.hal
+	$(transform-generated-source)
+LOCAL_GENERATED_SOURCES += $(GEN)
+include $(BUILD_STATIC_JAVA_LIBRARY)
+
+
+
+include $(call all-makefiles-under,$(LOCAL_PATH))
diff --git a/rgbled/1.0/IRgbLed.hal b/rgbled/1.0/IRgbLed.hal
new file mode 100644
index 0000000..d538603
--- /dev/null
+++ b/rgbled/1.0/IRgbLed.hal
@@ -0,0 +1,48 @@
+/*
+ * Copyright (C) 2026 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package vendor.lineage.rgbled@1.0;
+
+/**
+ * Direct control of the notification RGB LED, including kernel LED triggers.
+ * Brightness values are 0-255; colors are 0xRRGGBB.
+ */
+interface IRgbLed {
+    /** Set one channel's brightness; 0 turns it off. Other channels are untouched. */
+    setChannel(Channel channel, uint8_t brightness) generates (Status status);
+
+    /** Show a solid color, each component scaled by brightness. */
+    setColor(uint32_t rgb, uint8_t brightness) generates (Status status);
+
+    /** Blink a color using the PMIC hardware pattern generator. */
+    setBlink(uint32_t rgb, uint8_t brightness, uint32_t onMs, uint32_t offMs)
+            generates (Status status);
+
+    /** Kernel LED triggers available for the RGB channels (e.g. "heartbeat"). */
+    getSupportedTriggers() generates (vec<string> triggers);
+
+    /** Attach a kernel trigger to a channel; "none" detaches. */
+    setTrigger(Channel channel, string trigger) generates (Status status);
+
+    /** Currently active trigger of a channel. */
+    getTrigger(Channel channel) generates (string trigger);
+
+    /** Current brightness of a channel, or -1 on error. */
+    getBrightness(Channel channel) generates (int32_t brightness);
+
+    /** Detach triggers and switch all channels off. */
+    turnOff() generates (Status status);
+};
diff --git a/rgbled/1.0/types.hal b/rgbled/1.0/types.hal
new file mode 100644
index 0000000..bc2260b
--- /dev/null
+++ b/rgbled/1.0/types.hal
@@ -0,0 +1,31 @@
+/*
+ * Copyright (C) 2026 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package vendor.lineage.rgbled@1.0;
+
+enum Status : int32_t {
+    SUCCESS,
+    BAD_VALUE,
+    NOT_SUPPORTED,
+    IO_ERROR,
+};
+
+/** One physical channel of the notification RGB LED. */
+enum Channel : int32_t {
+    RED = 0,
+    GREEN = 1,
+    BLUE = 2,
+};
diff --git a/rgbled/Android.bp b/rgbled/Android.bp
new file mode 100644
index 0000000..bbb3e4b
--- /dev/null
+++ b/rgbled/Android.bp
@@ -0,0 +1,4 @@
+// This is an autogenerated file, do not edit.
+subdirs = [
+    "1.0",
+]
